prompting/apparmor/message.go

```diff
@@ -0,0 +1,436 @@
+func (msg *MsgNotification) UnmarshalBinary(data []byte) error {
+	const prefix = "cannot unmarshal apparmor notification message"
+
+	// Unpack fixed-size elements.
+	buf := bytes.NewBuffer(data)
+	order := arch.Endian() // ioctl messages are native byte order, but others might not be
+	if err := binary.Read(buf, order, msg); err != nil {
+		return fmt.Errorf("%s: cannot unpack: %s", prefix, err)
+	}
+
+	return nil
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotification) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, but others might not be
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// Validate returns an error if the mssage contains invalid data.
+func (msg *MsgNotification) Validate() error {
+	if !msg.NotificationType.IsValid() {
+		return fmt.Errorf("unsupported notification type: %d", msg.NotificationType)
+	}
+	return nil
+}
+
+// MsgNotificationUpdate (TBD, document me)
+//
+//	struct apparmor_notif_update {
+//	  struct apparmor_notif base;
+//	  __u16 ttl;          /* max keep alives left */
+//	} __attribute__((packed));
+type MsgNotificationUpdate struct {
+	MsgNotification
+	TTL uint16
+}
+
+// MsgNotificationResponse describes a response to a MsgNotification.
+//
+// This structure corresponds to the kernel type struct apparmor_notif
+// described below.
+//
+//	struct apparmor_notif_resp {
+//	  struct apparmor_notif base;
+//	  __s32 error;        /* error if unchanged */
+//	  __u32 allow;
+//	  __u32 deny;
+//	} __attribute__((packed));
+type MsgNotificationResponse struct {
+	MsgNotification
+	// XXX: The embedded MsgNotification also has an Error field, why?
+	Error int32
+	// Allow somehow encodes the allowed operation mask.
+	Allow uint32
+	// Deny somehow encodes the denied operation mask.
+	Deny uint32
+}
+
+// ResponseForRequest returns a response message for a given request.
+func ResponseForRequest(req *MsgNotification) MsgNotificationResponse {
+	return MsgNotificationResponse{
+		MsgNotification: MsgNotification{
+			NotificationType: APPARMOR_NOTIF_RESP,
+			// XXX: should Signalled be copied?
+			Signalled: req.Signalled,
+			// XXX: should Flags be copied?
+			Flags: req.Flags,
+			ID:    req.ID,
+			// XXX: should Error be copied?
+			Error: req.Error,
+		},
+	}
+}
+
+// MarshalBinary marshals the message into binary form.
+func (msg *MsgNotificationResponse) MarshalBinary() ([]byte, error) {
+	msg.Version = 2
+	msg.Length = uint16(binary.Size(*msg))
+	buf := bytes.NewBuffer(make([]byte, 0, msg.Length))
+	order := arch.Endian() // ioctl messages are native byte order, but others might not be
+	if err := binary.Write(buf, order, msg); err != nil {
+		return nil, err
+	}
+	return buf.Bytes(), nil
+}
+
+// MsgNotificationOp (TBD, document me).
```

Comment:
(nitpick) this should be lowercase as the msgNotificationOp - also I wonder if it should be `MsgNotificationOpKernel` or`MsgNotificationOpRaw` or something (like above?)